### DISCUSSION TITLE
Left side wave extension

### DISCUSSION THREAD

**Discussion by spamatica** [muse-sequencer/muse#1211] [Team Posts]
Hi Tim,
I saw you push your fixes now, nice! So I thought I would take it for a spin before finalizing the release.
I did however found a regression as the title says. If I have a wave file that I have been cutting in it is no longer possible to extend a wave part to the left.
Or rather the part is extended but the extended bit is empty. 
Could you take a look?
I don't know how extensive your changes are but I can try and have a peek too.

**Comment by spamatica** [comment]
Moving the left side to the right doesn't behave the same as before either. The wave is moved instead of being hidden.
Edit: This goes for midi parts also.

**Comment by terminator356** [comment]
Hi! I'm here, I'm awake, I'm watching.
 
It is no longer possible to have events hidden or partially hidden before the left border.
We could not support that feature at this time, massive work is needed.
Attempting to do so will push the events to the right so that nothing is hidden.

However, as a compromise of sorts, when dragging the borders, press Ctrl
 to drag all the events along with the border.

**Comment by spamatica** [comment]
Oh, hey there :) 

But it did work. Was it not possible to keep that? I don't remember what the issue was that you fixed..

**Comment by terminator356** [comment]
I explained it in a recent post in detail.
At first I was supportive of it. But then I completely reversed that stand.
Because we would need to change the fundamental unit of time from 32-bit unsigned integers
 to 64-bit signed integers.

Yes, it appeared to work, but under the hood it was not correct.
For example the event lists are sorted by unsigned integer time value.
That would need to change to signed. And a whole lot of other things.

**Comment by spamatica** [comment]
Yeah, I recall the mails. It appears I didn't grasp the consequences. We need some other way to solve this then, that feature was very useful.
For midi parts it does not matter as much as there are other ways to get by, but for audio there isn't.

After scratching my head a bit I think I remember my original idea, which I likely never got around to since I got it to work by having negative events. It should still work.
Basically, for wave parts, if an event starts on the same pos as the part (should be zero if my understanding is correct), the event should be kept at the start of the part and the spos should be adjusted to keep the wave "syncronized".
I can have a look to see if I can figure out where this change would go.

**Comment by terminator356** [comment]
Yes I saw exactly what you were trying to do.
Dynamically extend the first or last wave event when adjusting part borders.
But this cannot work.
We cannot be manipulating events via graphical operations in the Arranger.
The Arranger is for manipulating parts, not the events within them.
That is what the editors are for.
If wave events are to be adjusted, then open the corresponding editor and do it there.
To that end, that is why I now draw clear boundaries on the wave events in the Arranger,
 to show exactly the boundaries, and if they need adjusting then open the editor to do it.

**Comment by terminator356** [comment]
That said, yes we could use something in the Wave Editor to manipulate a wave's 'spos'.
Believe it or not I actually have code in there from long ago that does that.
I just never linked it up with an actual graphical control.
I did at first, but then I realised there were problems with adjusting that spos value
 so I removed the ability to adjust it.
So, now I think maybe I could (should) re-enable it somehow because we do need that ability.

**Comment by terminator356** [comment]
I was going to try not only an 'spos' numerical adjustment box but also the ability
 to manipulate 'spos' by dragging wave event borders while holding a key down.

**Comment by spamatica** [comment]
Are you saying to not do what I proposed on the account of we should not manipulate events in the arranger?
This seems very constricting without a good reason. What I had implemented may have worked through a glitch but it is exactly the right way to adjust wave parts. And it is the way it works in for instance Cubase.
Besides, that wave parts have events isn't really something the user is aware of it's more of an implementation detail.
Or are you saying that it cannot work to do it by manipulating the event in the arranger?

**Comment by terminator356** [comment]
But as I mentioned before the way it was done was causing some problems.

That wave parts have events is what I am trying to make clear to the user.
Yes it's tough when there's only one wave event, it's hard to make it clear.
But once users see the power of 'gluing' several wave parts together, they'll see them as
 events within the part.
And that's when this auto-extending feature breaks down. It arbitrarily alters events
 when the user might not have wanted it. Why not do the same for midi events then?

**Comment by terminator356** [comment]
If you give me a few days I promise to try and make the hidden events at left left border thing work.
Possibly by tonight.
I'm just saying it's not correct and there will be problems down the road, especially if people try
 to add features that support it.

I am starting right now.
Please let me handle it and commit?

**Comment by spamatica** [comment]
I would appreciate that.
I see what you are going for with exposing events so it's evident that
parts have been glued together, but I still defend that the event at pos
zero is special and it is logical that it would reveal more wave, if it
exists.

Btw, regarding glue I noticed that gluing waves still caused doubling of
the audio when I tried.

Cheers

Den sön 4 apr. 2021 21:58Tim ***@***.***> skrev:

**Comment by terminator356** [comment]
If I can make it work decently, I believe it may accomplish what you were trying to do,
 which I believe was to simply allow left border resizing so that (wave) parts could
 have a 'window' onto their content.
If this works, no events at all will need modification.
I too was saddened that I had to throw out the left side resizing. It would have been so cool,
 this 'windowing' scheme. So yeah I am motivated to make it work.

I have it working right now with full left and right border sizing, and left and right
 'hidden events' markers working.
Some things needed changing. For example loading and saving didn't like negative values.

The midi is working just fine at the moment.
Still there are a couple of things with the waves to be checked and fixed.
Hang in there...

**Comment by terminator356** [comment]
Breakthrough! This screenshot shows a wave part that has been split into two and a midi part.
All have had their borders adjusted. You can see the left and right 'hidden events' arrows are active.
Note that the right wave part has an intrinsic offset (the wave event's 'spos' member) due to the split
 and this is automatically taken into account by the low level wave class.

When I play this, it works.

All the the waves are properly offset, and sound in sync with what's on the screen.
**Most importantly**, when the wave parts are placed at bar one, that is when the wave events
 (which are hidden before the parts' left borders) have **negative** absolute time values.
This was crucial to get right since, when the parts are placed some distance into the song
 the absolute time values of the wave events are still not negative yet although they are negative
 with respect to the parts left edge.

It's working quite splendidly, I must say, all the tests pass so far, loading, saving,
 various part border sizes and placements etc. Man, it took a few night to get one
 crucial statement correct, where the wave's seek offset is calculated, so that it
 starts and plays correctly no matter where the play-head is placed.

There are still some graphics things to fix, very similar to these sound fixes.
Please hang in there, I'll try to ensure the thing is usable.

![left border fixes 1](https://user-images.githubusercontent.com/5422362/113968275-c3f57f00-9800-11eb-8502-1fd389ecb73e.png)

**Comment by spamatica** [comment]
Alright, that looks great!

I'm grateful for you working on this.

/Robert

Den tors 8 apr. 2021 kl 06:39 skrev Tim ***@***.***>:

**Comment by terminator356** [comment]
Hope you're still awake.
Check it out!
See the ChangeLog or commit to peruse what I changed.

As I stress in the log, it should be fine and any new bugs should only be with hidden left border events,
 hopefully...
```
      ANY BUGS SHOULD only occur when there are left border hidden events, otherwise
       it SHOULD behave normally.
      Should be OK to use but still some areas need attention such as 'functions',
       merging and splitting, and more.
```

**Comment by terminator356** [comment]
PS: I see that looping waves is broken recently even without latency compensation,
 and it has been reported by users.
Hmm, I don't recall changing anything in there. I recall some recent looping changes by kybos.
I'll try to take a look, not sure how long it would take to fix at this point, have not investigated yet.

**Comment by spamatica** [comment]
Looks great in my initial test!
I'm grateful for this fix, and a little bit guilty for stressing it...

Regarding the looping, it seems you are right. I nearly always use latency
correction so I've grown out of looping so I didn't notice.
Yeah, seems like we should probably fix that before the release. At least
wait til tomorrow and kybos can comment if there is any probable cause?

/Robert

Den sön 11 apr. 2021 kl 21:17 skrev Tim ***@***.***>:

**Comment by terminator356** [comment]
Checking... I see why it's not looping. But not what caused it to break.
I might have a fix soon-ish.
Not sure what caused that. Things like that are usually my fault.

**Comment by terminator356** [comment]
Not as easy as I suspected. Looks like this may take some time to fix.
I think I broke it a little over a year ago when I tried to do some dynamic pre-fetch audio cache stuff.
It had some bugs so I have a branch where I was trying to fix it, but the fix didn't work out well,
 so I was trying to think of a better fix.
Then things fell apart. Time wise, pandemic wise, work wise. Sadly I've forgotten some of what I was doing.
I must re-learn what the problem was and what I was trying to do, then re-learn what I thought
 the better fix might be... This part of the code was really complex.

**Comment by spamatica** [comment]
Oh well. Not ideal but maybe we do the release anyway then and aim for a
4.01 when it is possible?

Den mån 12 apr. 2021 kl 07:21 skrev Tim ***@***.***>:

**Comment by kybos** [comment]
I'm quite certain I haven't touched anything affecting looping the waves (nor the wave editor internals in general, for that matter...).

**Comment by spamatica** [comment]
I don't think it's related to the wave editor, though as Tim wrote above it has likely been broken since his changes about a year ago.
Looping does still work with midi tracks so there is that.
Unless you guys say otherwise I think I will go ahead and do the release anyway, tomorrow night if things go well.

**Comment by terminator356** [comment]
Yeah my fault.
I have run diagnostics and crunched the numbers and I see where the problem is
 and I remember what I was trying to do. I am now working to fix it.
Kind of a bummer releasing 4.0 with this bug.
Heh, you can just bet that looping is the first thing many will try and the complaints will start rolling in...
If you would like to wait a few days to see what I can do, maybe pull another miracle from my hat.
Either way I promise to keep on this and try to fix this ASAP, 4.0 released or not.

**Comment by kybos** [comment]
I must agree with Tim. I think it's not a good idea to do a major release with basic bugs like that. I guess the new release will raise some interest, people will try it out, but when they run in such bugs, the reactions could be very negative.
On the other side, we can't go on like that indefinitely. What about a deadline next week on the weekend? Hoping that Tim can fix it by then and Robert will have time to do the release?

**Comment by spamatica** [comment]
Den ons 14 apr. 2021 08:27kybos ***@***.***> skrev:

> I must agree with Tim. I think it's not a good idea to do a major release
> with basic bugs like that. I guess the new release will raise some
> interest, people will try it out, but when they run in such bugs, the
> reactions could be very negative.
> On the other side, we can't go on like that indefinitely. What about a
> deadline next week on the weekend? Hoping that Tim can fix it by then and
> Robert will have time to do the release?
>

Alright, good to know the general feeling. Sounds reasonable, I'll check
again second half next week.

A related matter if we, as you say, hope for raised interest.
Apart from uploading the release, updating our news pages and posting in
Linuxmusicians forums, should we include some other channel? Possibly make
some kind of event?
New videos would be great also.

Any ideas?

**Comment by kybos** [comment]
> A related matter if we, as you say, hope for raised interest.
> Apart from uploading the release, updating our news pages and posting in
> Linuxmusicians forums, should we include some other channel? Possibly make
> some kind of event?
> New videos would be great also.
> 
> Any ideas?

What comes to mind:
- Revamp/Extend/Update the Wikipedia article.
- Update the screenshots on MusE Web (I already transferred the existing screenshots to the corresponding wiki page from there, to keep the history).
- Update the wiki documentation, at least the (mostly very old) screenshots.
- Maybe inform some of the Linux/FOSS web magazines that typically review/present new software (?).

**Comment by terminator356** [comment]
You know, I could swear the looping was actually working.
There was only an issue with looping while latency correction was on.
Do you remember? It was fine. That's why I left it alone. I could swear something changed !
My branch (in the git tree) was titled 'latency_looping' and now that I have loaded it up again, I see
 that it was an attempt to fix just that - to finally fix looping while latency correction was on.
(Actually I see that I went **much** further than that. I was attempting to also correct several other things
 that are sensitive to latency, such as the playback of audio controller graphs.)

Unfortunately, due to deprecated Qt code and so on, I must merge the latest master code to my branch.
And that's where big trouble is happening.
Due to the move from 'muse3' folder to 'src' folder, it refuses to merge properly.
I fear I must now merge all my changes by hand.
I wish there was a way to get a diff against what the master was **then**, not **now**,
 so that I could more easily manually apply the diffs against current master.
So far everything I've tried fails.
(The reason that the merge worked well on my last commits is that those local works in progress
 were based upon the master, they were not separate branches like this branch 'latency_looping'.)
I'll keep trying...

**Comment by terminator356** [comment]
**Eek! False alarm!** It **IS** working. With Jack only.
Everything is working as it was. The dummy that I am, I was testing with the **dummy** driver ! 
Terribly sorry for the noise (or lack of it in this case) and the delay, bro.
You may now release at will, sir.

Technical: Jack has a transport mechanism. When our 'use Jack transport' button is **off** or
 another non-Jack driver is used, our internal pseudo-transport kicks in to keep things rolling.
I wrote it to imitate Jack's transport as best as possible.
However, one small detail is not exactly the same as Jack: The number of cycles between
 calling a Jack transport relocation (for example at loop time) and it actually taking effect
 is I believe two cycles. But I think my internal pseudo-transport does it in one cycle.
So the looping code becomes confused in that scenario. Well, something like that...

So long story short, use Jack (and no latency correction) and it will be OK.

**Comment by kybos** [comment]
Good news!

But is also means that it will never work without Jack? If this is the case, it should be clearly stated and documented IMO to avoid false expectations.
Currently the documentation says that PulseAudio is also supported, which suggests feature parity:

> MusE now supports PulseAudio, in this scenario you don't even need jack. For realtime playing, the latency will likely be an issue though, i personally use it all the time while mixing as it is very fast to setup.
> 
> Type muse3 -t to start MusE with pulse audio backend.

And there is an open issue on it as well, declaring it as a bug:
https://github.com/muse-sequencer/muse/issues/806

**Comment by terminator356** [comment]
Well, not 'never'. I just need to correct my internal transport to mimic the number of cycles that Jack uses.
It **will** work and **can** be fixed. I looked at the code last night and I'm working on it, although no quick fix yet.
But using a quick and dirty fix I proved that it will work. I had it working last night.

Actually I am much more interested in getting my much more complete latency_looping branch to run,
 because it finally fixes the broken looping with latency correction turned on.
I am very intrigued to get it going because now I think I have spotted a reason why it didn't work.

**Comment by terminator356** [comment]
> It will work and can be fixed

Done. Looping: Fixed silent wave tracks when Jack transport off or non-Jack driver used.
Tested OK with the following drivers and situations:
1) Jack and 'Use Jack Transport' button off
2) RTAudio: Pulse, ALSA etc.
4) Dummy driver

RTAudio support could ever really have 'feature parity' with Jack, because RTAudio is a wrapper -
 it can't support every feature of all the servers. Last time I checked it has no routing, no latency reporting
 (thus we can't correct it), no freewheel mode. Jack has those. Even if RTAudio did have those things,
 each server might not, such as Pulse.

So... Now to concentrate on the latency looping bug, in my branch.
To recap, this bug is: With Jack and latency correction enabled, looping waves doesn't work correctly.

The app's looking sharp, kybos. You're a hell of a trooper for sticking with it. Stunning transformation, really.
If I ever get all the engine things I had on the go done, it will really rock.

**Comment by terminator356** [comment]
A final thought tonight, Probably more important than the remaining latency looping bug is the
 small segment of missing playback or record content when latency correction is on.
I know Robert hates it. Yeah, I'm workin' on it planning wise, I've got two paths to try.
I wish I had more time these days to implement it.

**Comment by kybos** [comment]
> Done. Looping: Fixed silent wave tracks when Jack transport off or non-Jack driver used.
> Tested OK with the following drivers and situations:
> 
>     Jack and 'Use Jack Transport' button off
>     RTAudio: Pulse, ALSA etc.
>     Dummy driver

That's good news again, thanks! I've closed the ticket.

> RTAudio support could ever really have 'feature parity' with Jack, because RTAudio is a wrapper -
> it can't support every feature of all the servers. Last time I checked it has no routing, no latency reporting
> (thus we can't correct it), no freewheel mode. Jack has those. Even if RTAudio did have those things,
> each server might not, such as Pulse.

Sure. It's about the basic functionality (like the looping). 
The missing latency correction with Pulse is mentioned in the docu indeed, but not clearly enough as a restriction of RTAudio/Pulse, it sounds more like a shortcoming of MusE. I will change that and try to make it clearer.
 
> The app's looking sharp, kybos. You're a hell of a trooper for sticking with it. Stunning transformation, really.
> If I ever get all the engine things I had on the go done, it will really rock.

Thanks. I'm still convinced that there is a lot of potential in MusE (even if the "competition" is getting stronger, too).

**Comment by spamatica** [comment]
Hey guys, wow, reading up on the progress and this is looking good :) No time tomorrow but I will work on the release on Sunday. Including some of the work Kybos suggested above, updating screenshots etc, I don't think I will get it done but it's closing in :)

> Due to the move from 'muse3' folder to 'src' folder, it refuses to merge properly.
> I fear I must now merge all my changes by hand.
> I wish there was a way to get a diff against what the master was then, not now,

Not sure what you tried but there must be a way to get git to help you.
Did you try merging to a checkout before the src switch? 
Maybe create a merge branch just before the src change and merge into that one and then merge that one to master, making it a two step process? In my mind that should work :) Not that I'm that much of a git wizard, there might be other approaches.

Regarding feature parity of RTAudio, it is true that some features are simply not possible to achieve. But freewheeling should still be possible to implement. Apart from simply not sending any audio to RTAudio I guess it would require the internal transport you mentioned to allow for freewheeling.
On that note, I have probably mentioned this before, having a dedicated Alsa backend would be a good thing also for those like me that want to cram every last bit of performance out of the hardware.

**Comment by spamatica** [comment]
> But freewheeling should still be possible to implement. Apart from simply not sending any audio to RTAudio I guess it would require the internal transport you mentioned to allow for freewheeling.

So, I got inspired, and maybe to prove my point, I went ahead and tried to implement this. Of course it turns out I got it partly backwards, but still it was pretty easy to get something working.
No mutexes or anything this far, not sure any are needed? I think they might... in any case I checked it into a branch (rtaudio_freewheel), a bit late to include it without extensive testing.
It is something I have needed though, as it happens pretty often that I do downmixes while using the rtaudio backend and it can be kind of slow.

**Comment by terminator356** [comment]
Wow, that's inspired alright. Roll up the ol' sleeves and try.
I suppose it should be possible. And our internal transport is ready and waiting.
I'll check it out and think about how I might do such a thing, maybe offer some pointers.
Keep in mind the dummy driver as well. Some users might do quick mixdowns with the dummy driver.
Thus the freewheel could be universal regardless of driver.

Hm, why do you use the RTAudio driver? Seems a bit odd.

At some point we should add automatically restarting MusE when users change the driver, or anything
 else requiring a restart, since we have that restart mechanism in place for the Appearance dialog.
It would be cool if we could switch drivers without restarting. Briefly thought about it a few times...

**Comment by terminator356** [comment]
About a native ALSA driver: Old MusE had one but it was removed. Could be a place to start.
I have the code around here.

Notice that some heavyweights are leaning towards ALSA instead of Jack now.
Ardour tends to favour it. I often follow what Mr. Davis says, and his position seems to have changed
 over the years. More in favour of a monolithic 'do-everything' DAW these days, meaning you can
 get away with using the ALSA system in exclusive mode for all the features (no other apps can open it).

**Comment by spamatica** [comment]
> Keep in mind the dummy driver as well. Some users might do quick mixdowns with the dummy driver.
> Thus the freewheel could be universal regardless of driver.

Hmm, could you? I mean... I guess it's technically possible but you are completely in the dark as you can't hear what you are doing prior to the mixdown. Still it would be easy to implement for the dummy driver so it doesn't matter :)
 
> Hm, why do you use the RTAudio driver? Seems a bit odd.

Jack does make some things harder, like listening to reference material during mixing. And for whatever reason it looks like I can get more stable playback under very high cpu when using the pulse backend. Thinking about it this is probably mostly because the pulse backend tends to have very high latency.

> Notice that some heavyweights are leaning towards ALSA instead of Jack now.

Yeah, jack was a very interesting idea at the time but it turns out I personally hardly ever utilize the multi application connectivity, and probably very few people does. Going directly for the bare metal, alsa, should allow a little bit lower latency as less code is executed in the realtime scope.
It's not a super high priority in my book, but some day maybe :)
